subsys/bluetooth/mesh/light_ctrl_regulator.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bluetooth/mesh/light_ctrl_regulator.h>
+#include <bluetooth/mesh/light_ctrl_srv.h>
+
+#define REG_INT CONFIG_BT_MESH_LIGHT_CTRL_SRV_REG_INTERVAL
+
+static void reg_step(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_mesh_light_ctrl_spec_regulator *spec_reg = CONTAINER_OF(
+		dwork, struct bt_mesh_light_ctrl_spec_regulator, timer);
+
+	if (!bt_mesh_light_ctrl_srv_is_enabled(spec_reg->srv)) {
+		/* The server might be disabled asynchronously. */
+		return;
+	}
+
+	k_work_reschedule(&spec_reg->timer, K_MSEC(REG_INT));
+
+	float target = bt_mesh_light_ctrl_srv_reg_target_get(spec_reg->srv);
+	float ambient = bt_mesh_light_ctrl_srv_reg_ambient_get(spec_reg->srv);
+	float error = target - ambient;
+	float accuracy = bt_mesh_light_ctrl_srv_reg_accuracy_get(spec_reg->srv, target);
+	float input;
+
+	if (error > accuracy) {
+		input = error - accuracy;
+	} else if (error < -accuracy) {
+		input = error + accuracy;
+	} else {
+		input = 0.0f;
+	}
+
+	float kp, ki;
+
+	if (input >= 0) {
+		bt_mesh_light_ctrl_srv_reg_coeff_up_get(spec_reg->srv, &kp, &ki);
+	} else {
+		bt_mesh_light_ctrl_srv_reg_coeff_down_get(spec_reg->srv, &kp, &ki);
+	}
+
+	spec_reg->i += (input * ki) * ((float)REG_INT / (float)MSEC_PER_SEC);
+	spec_reg->i = CLAMP(spec_reg->i, 0, UINT16_MAX);
+
+	float p = input * kp;
+	uint16_t output = CLAMP(spec_reg->i + p, 0, UINT16_MAX);
+
+	/* The regulator output is always in linear format. We'll convert to
+	 * the configured representation again before calling the Lightness
+	 * server.
+	 */
+	uint16_t lvl = bt_mesh_light_ctrl_srv_configured_light_get(spec_reg->srv);
+
+	/* Output value is max out of regulator and configured level. */
+	if (output > lvl) {
+		if (output == spec_reg->prev) {
+			return;
+		}
+
+		atomic_set(&spec_reg->flag, true);
+	} else if (atomic_set(&spec_reg->flag, false)) {
+		output = lvl;
+	} else {
+		return;
+	}
+
+	spec_reg->prev = output;
+
+	bt_mesh_light_ctrl_srv_reg_update(spec_reg->srv, output);
+}
+
+static void reg_init(struct bt_mesh_light_ctrl_srv *srv, void *user_data)
+{
+	struct bt_mesh_light_ctrl_spec_regulator *spec_reg =
+		(struct bt_mesh_light_ctrl_spec_regulator *) user_data;
+	k_work_init_delayable(&spec_reg->timer, reg_step);
+	spec_reg->srv = srv;
+}
+
+static void reg_start(struct bt_mesh_light_ctrl_srv *srv, void *user_data)
+{
+	struct bt_mesh_light_ctrl_spec_regulator *spec_reg =
+		(struct bt_mesh_light_ctrl_spec_regulator *) user_data;
+	k_work_schedule(&spec_reg->timer, K_MSEC(REG_INT));
+}
+
+static void reg_stop(struct bt_mesh_light_ctrl_srv *srv, void *user_data)
+{
+	struct bt_mesh_light_ctrl_spec_regulator *spec_reg =
+		(struct bt_mesh_light_ctrl_spec_regulator *) user_data;
+	spec_reg->i = 0;
+	spec_reg->prev = 0;
+	k_work_cancel_delayable(&spec_reg->timer);
```

Comment:
Note that cancel can fail, and that the [workqueue best practices](http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/reference/kernel/threads/workqueue.html#check-return-values) documentation suggests to document ignored return codes for this stating why this can be safely ignored.